File: operator/pkg/controllers/resource.go
/*
Copyright 2022.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package controllers

import (
	"fmt"
	"context"
	"github.com/golang/glog"
	elasticv1alpha1 "github.com/intelligent-machine-learning/easydl/operator/api/v1alpha1"
	commonv1 "github.com/intelligent-machine-learning/easydl/operator/pkg/common/api/v1"
	common "github.com/intelligent-machine-learning/easydl/operator/pkg/common"
	logger "github.com/sirupsen/logrus"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	utilpointer "k8s.io/utils/pointer"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"k8s.io/apimachinery/pkg/api/errors"
)

const (
	labelAppName     = "app"
	labelJobName     = "elasticjob-name"
	easydlApp        = "easydl"
	envMasterAddrKey = "MASTER_ADDR"
)

// PodManager manages the lifecycle of a pod including creation, updation and deletion.
type PodManager struct{}

func newPodManager() *PodManager {
	return &PodManager{}
}

// NewPod creates a Pod according to a PodTemplateSpec
func (m *PodManager) NewPod(
	job *elasticv1alpha1.ElasticJob,
	podTemplate *corev1.PodTemplateSpec,
	podName string,
) *corev1.Pod {
	podSpec := podTemplate.DeepCopy()

	if len(podSpec.Labels) == 0 {
		podSpec.Labels = make(map[string]string)
	}

	if len(podSpec.Annotations) == 0 {
		podSpec.Annotations = make(map[string]string)
	}
	podSpec.Labels[labelAppName] = easydlApp
	podSpec.Labels[labelJobName] = job.Name

	for key, value := range job.Labels {
		podSpec.Labels[key] = value
	}

	for key, value := range job.Annotations {
		podSpec.Annotations[key] = value
	}

	if len(podSpec.Spec.Containers) == 0 {
		glog.Errorf("Pod %s does not have any container", podName)
		return nil
	}

	setMasterAddrIntoContainer(&podSpec.Spec.Containers[0], job.Name)
	pod := &corev1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:        podName,
			Namespace:   job.Namespace,
			Labels:      podSpec.Labels,
			Annotations: podSpec.Annotations,
			OwnerReferences: []metav1.OwnerReference{
				*metav1.NewControllerRef(job, elasticv1alpha1.SchemeGroupVersionKind),
			},
		},
		Spec: podSpec.Spec,
	}
	return pod
}

// DeletePod remove a Pod 
func (m *PodManager) DeletePod(
	r *ElasticJobReconciler,
	job *elasticv1alpha1.ElasticJob,
	pod *corev1.Pod,
) {
	deleteOptions := &client.DeleteOptions{GracePeriodSeconds: utilpointer.Int64Ptr(0)}
	if err := r.Delete(context.Background(), pod, deleteOptions); err != nil && !errors.IsNotFound(err) {
		r.Recorder.Eventf(
			job,
			corev1.EventTypeWarning,
			common.JobScalingReason,
			"Error deleting worker %s: %v",
			pod.Name,
			err,
		)
	}
}

// GetReplicaTypePods get all ReplicaType Pods of a job
func (m *PodManager) GetReplicaTypePods(
	r *ElasticJobReconciler,
	job *elasticv1alpha1.ElasticJob,
	replicaType commonv1.ReplicaType,
) ([]corev1.Pod, error) {
	replicaLabels := make(map[string]string)
	replicaLabels[LabelReplicaTypeKey] = string(replicaType)
	replicaLabels[labelJobName] = job.Name
	labelSelector := &metav1.LabelSelector{
		MatchLabels: replicaLabels,
	}
	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
	if err != nil {
		logger.Warningf("No selector found")
	}
	podlist := &corev1.PodList{}
	err = r.List(context.Background(), podlist, client.MatchingLabelsSelector{Selector: selector})
	if err != nil {
		return nil, err
	}
	return podlist.Items, nil
}

// GetReplicaStatus gets ReplicaStatus from ReplicaType Pods
func (m *PodManager) GetReplicaStatus(pods []corev1.Pod) *commonv1.ReplicaStatus {
	replicaStatus := commonv1.ReplicaStatus{}

	for _, pod := range pods {
		if pod.Status.Phase == corev1.PodPending {
			replicaStatus.Pending++
		} else if pod.Status.Phase == corev1.PodRunning {
			replicaStatus.Active++
		} else if pod.Status.Phase == corev1.PodFailed {
			replicaStatus.Failed++
		} else if pod.Status.Phase == corev1.PodSucceeded {
			replicaStatus.Succeeded++
		}
	}
	return &replicaStatus
}

// NewService create a service
func (m *PodManager) NewService(job *elasticv1alpha1.ElasticJob, name string, port int32, selector map[string]string) *corev1.Service {
	selector[labelAppName] = easydlApp
	selector[labelJobName] = job.Name
	return &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      name,
			Namespace: job.Namespace,
			Labels:    selector,
			OwnerReferences: []metav1.OwnerReference{
				*metav1.NewControllerRef(job, elasticv1alpha1.SchemeGroupVersionKind),
			},
		},
		Spec: corev1.ServiceSpec{
			ClusterIP: "None",
			Selector:  selector,
			Ports: []corev1.ServicePort{
				{
					Port: port,
				},
			},
		},
	}
}

func setMasterAddrIntoContainer(container *corev1.Container, jobName string) {
	masterAddrEnv := newMasterAddrEnvVar(jobName)
	container.Env = append(container.Env, masterAddrEnv)
}

func newMasterAddrEnvVar(jobName string) corev1.EnvVar {
	masterServiceAddr := NewEasydlMasterName(jobName)
	return corev1.EnvVar{
		Name:  envMasterAddrKey,
		Value: fmt.Sprintf("%s:%d", masterServiceAddr, masterServicePort),
	}
}

File: dlrover/python/common/resource.py
# Copyright 2025 The EasyDL Authors. All rights reserved.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from typing import Dict


class Resource(object):
    def __init__(self, cpu=0, memory=0, disk=0, gpu=0, ud_resource=None):
        self._cpu: float = cpu
        self._memory: int = memory
        self._disk: int = disk
        self._gpu: float = gpu
        self._gpu_type: str = ""

        self._ud_resource: Dict[str, float] = ud_resource

    def __repr__(self):
        return (
            f"Resource(cpu={self._cpu}, "
            f"mem={self._memory}, "
            f"disk={self._disk}, "
            f"gpu={self._gpu}, "
            f"gpu_type={self._gpu_type}, "
            f"user_defined={self._ud_resource})"
        )

    @property
    def cpu(self):
        return self._cpu

    @property
    def memory(self):
        return self._memory

    @property
    def disk(self):
        return self._disk

    @property
    def gpu(self):
        return self._gpu

    @property
    def gpu_type(self):
        return self._gpu_type

    @property
    def ud_resource(self):
        return self._ud_resource

    @classmethod
    def default_cpu(cls):
        return Resource(cpu=1)

    @classmethod
    def default_gpu(cls):
        return Resource(gpu=1)

    @classmethod
    def from_dict(cls, resource_dict):
        if not resource_dict:
            return Resource()

        cpu_value = 0
        mem_value = 0
        disk_value = 0
        gpu_value = 0
        ud_value = {}

        for key, value in resource_dict.items():
            if key.lower() == "cpu":
                cpu_value = value
            elif key.lower() == "memory" or key.lower() == "mem":
                mem_value = value
            elif key.lower() == "disk":
                disk_value = value
            elif key.lower() == "gpu":
                gpu_value = value
            else:
                ud_value[key] = value

        return cls(
            cpu=cpu_value,
            memory=mem_value,
            disk=disk_value,
            gpu=gpu_value,
            ud_resource=ud_value,
        )

    def to_dict(
        self,
        cpu_flag="cpu",
        gpu_flag="gpu",
        mem_flag="memory",
        disk_flag="disk",
    ) -> Dict[str, float]:
        result = {}
        if cpu_flag and self.cpu > 0:
            result[cpu_flag] = self.cpu
        if gpu_flag and self.gpu > 0:
            result[gpu_flag] = self.gpu
        if mem_flag and self.memory > 0:
            result[mem_flag] = self.memory
        if disk_flag and self.disk > 0:
            result[disk_flag] = self.disk

        return result

    def validate(self) -> bool:
        if self.cpu < 0 or self.memory < 0 or self.disk < 0 or self.gpu < 0:
            return False
        return True
